### ISSUE TITLE
Unit tests for providers (and template)

### ISSUE THREAD

**Issue by rugk** [libdns/libdns#44]
I've skimmed [across your implementation doc](https://github.com/libdns/libdns/wiki/Implementing-providers) and especially [your template repo](https://github.com/libdns/template), and I wonder whether some simple unit tests (included in the template) would not be a good idea?
Maybe also directly include a GitHub Action to run these tests in a CI.

As such, IMHO; this would make developing a new provider even easier by having a verification whether everything works (somewhat, at least) and support proper development practices.
This could make it easier to test the providers.

**Comment by mholt** [comment]
Yeah, that'd be a good idea, although, it's not super clear to me what "unit tests" would look like, since the output of each function is an external side effect on the network. We'd probably need mocks, which have to be customized for each provider... I hesitate to do DNS lookups since we've found they're extremely finicky in CI and people's various dev machines because people have the most convoluted, weirdest DNS resolver configurations that break tests like this...

Probably the most useful tests would be those which verify that API requests are being formed correctly, rather than testing the actual execution of those requests. I don't know.

**Comment by Adphi** [comment]
Conformance tests would work.
 
The main library implements the main conformance tests logic. This is what [Cert-Manager](https://github.com/cert-manager/cert-manager/blob/master/test/acme/dns/options.go) or the [CSI](https://github.com/kubernetes-csi/csi-test/tree/master/pkg/sanity) does.
The conformance tests only takes a `Provider` and each provider must ensure tests pass, so they validate it with their own development credentials.

**Comment by mholt** [comment]
I like that idea. It'd be awesome if it could somehow get each provider to use a mock DNS system somehow, operated by the conformance tests. But I don't know if that's feasible.

**Comment by Adphi** [comment]
While working with multiple implementations (scaleway, ovh, hetzner and cloudflare) what I found is that you can manage the zones in this providers without actually using their dns servers 😊. It just needs an account. I used one of my test domains for all of them but it may even work with anything, e.g. example.org

I think it should not use a mock at all as it would just validate the mock. Conformance tests should make real network calls with a test account and a test domain.

**Comment by Adphi** [comment]
I know it is off topic but some kind of `Capabilities` interface to know what kind of records are supported by the provider could be very useful / helpful.

**Comment by mholt** [comment]
Maybe, but I also kind of just expect DNS providers to support all the common record types, at least.

**Comment by Adphi** [comment]
Exactly, you'd expect it. With capabilities, you'd be sure of it at runtime before any attempt.

**Comment by mholt** [comment]
Let's talk about that in another issue perhaps

**Comment by gucci-on-fleek** [comment]
Discussed also in https://github.com/libdns/libdns/pull/153#issuecomment-2754133316.

**Comment by gucci-on-fleek** [comment]
See also libdns/libdns#189.
